File: generated_codes/chat_agents/chat-agents-cpp/skiplist/gpt5.5/output1_skiplist.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <random>

class ConcurrentDataStructure : public SetADT {
private:
    static constexpr int MAX_LEVEL = 16;

    struct Node {
        int val;
        int topLevel;
        std::atomic<Node*> forward[MAX_LEVEL];

        Node(int v, int level) : val(v), topLevel(level) {
            for (int i = 0; i < MAX_LEVEL; ++i) {
                forward[i].store(nullptr, std::memory_order_relaxed);
            }
        }
    };

    Node* head;
    Node* tail;

    static bool is_marked_ref(Node* p) {
        return (reinterpret_cast<uintptr_t>(p) & 1L) != 0;
    }

    static Node* get_unmarked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1L);
    }

    static Node* get_marked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1L);
    }

    int randomLevel() {
        thread_local std::mt19937 gen(std::random_device{}());
        thread_local std::uniform_int_distribution<int> dist(0, 1);

        int level = 0;
        while (level < MAX_LEVEL - 1 && dist(gen)) {
            ++level;
        }
        return level;
    }

    bool find(int key, Node* preds[], Node* succs[]) {
    retry:
        while (true) {
            Node* pred = head;

            for (int level = MAX_LEVEL - 1; level >= 0; --level) {
                Node* curr = get_unmarked_ref(
                    pred->forward[level].load(std::memory_order_acquire));

                while (true) {
                    Node* succ = curr->forward[level].load(std::memory_order_acquire);

                    while (is_marked_ref(succ)) {
                        Node* unmarkedSucc = get_unmarked_ref(succ);

                        if (!pred->forward[level].compare_exchange_strong(
                                curr,
                                unmarkedSucc,
                                std::memory_order_acq_rel,
                                std::memory_order_acquire)) {
                            goto retry;
                        }

                        curr = get_unmarked_ref(
                            pred->forward[level].load(std::memory_order_acquire));

                        succ = curr->forward[level].load(std::memory_order_acquire);
                    }

                    if (curr->val < key) {
                        pred = curr;
                        curr = get_unmarked_ref(succ);
                    } else {
                        break;
                    }
                }

                preds[level] = pred;
                succs[level] = curr;
            }

            return (succs[0]->val == key);
        }
    }

public:
    ConcurrentDataStructure() {
        head = new Node(INT_MIN, MAX_LEVEL - 1);
        tail = new Node(INT_MAX, MAX_LEVEL - 1);

        for (int i = 0; i < MAX_LEVEL; ++i) {
            head->forward[i].store(tail, std::memory_order_relaxed);
        }
    }

    ~ConcurrentDataStructure() override {
        Node* curr = head;

        while (curr != nullptr) {
            Node* next = nullptr;

            if (curr->val != INT_MAX) {
                next = get_unmarked_ref(
                    curr->forward[0].load(std::memory_order_acquire));
            }

            delete curr;

            if (curr->val == INT_MAX) {
                break;
            }

            curr = next;
        }
    }

    bool contains(int key) override {
        Node* curr = head;

        for (int level = MAX_LEVEL - 1; level >= 0; --level) {
            curr = get_unmarked_ref(curr);

            Node* next = get_unmarked_ref(
                curr->forward[level].load(std::memory_order_acquire));

            while (next->val < key) {
                curr = next;
                next = get_unmarked_ref(
                    curr->forward[level].load(std::memory_order_acquire));
            }
        }

        curr = get_unmarked_ref(
            curr->forward[0].load(std::memory_order_acquire));

        return curr->val == key &&
               !is_marked_ref(
                   curr->forward[0].load(std::memory_order_acquire));
    }

    bool add(int key) override {
        int topLevel = randomLevel();
        Node* preds[MAX_LEVEL];
        Node* succs[MAX_LEVEL];

        while (true) {
            bool found = find(key, preds, succs);

            if (found) {
                return false;
            }

            Node* newNode = new Node(key, topLevel);

            for (int level = 0; level <= topLevel; ++level) {
                newNode->forward[level].store(
                    succs[level],
                    std::memory_order_relaxed);
            }

            Node* expected = succs[0];

            if (!preds[0]->forward[0].compare_exchange_strong(
                    expected,
                    newNode,
                    std::memory_order_acq_rel,
                    std::memory_order_acquire)) {
                delete newNode;
                continue;
            }

            for (int level = 1; level <= topLevel; ++level) {
                while (true) {
                    expected = succs[level];

                    if (preds[level]->forward[level].compare_exchange_strong(
                            expected,
                            newNode,
                            std::memory_order_acq_rel,
                            std::memory_order_acquire)) {
                        break;
                    }

                    find(key, preds, succs);
                }
            }

            return true;
        }
    }

    bool remove(int key) override {
        Node* preds[MAX_LEVEL];
        Node* succs[MAX_LEVEL];
        Node* nodeToRemove = nullptr;
        bool isMarked = false;
        int topLevel = -1;

        while (true) {
            bool found = find(key, preds, succs);

            if (!found) {
                return false;
            }

            nodeToRemove = succs[0];
            topLevel = nodeToRemove->topLevel;

            for (int level = topLevel; level >= 1; --level) {
                Node* succ = nodeToRemove->forward[level].load(
                    std::memory_order_acquire);

                while (!is_marked_ref(succ)) {
                    nodeToRemove->forward[level].compare_exchange_strong(
                        succ,
                        get_marked_ref(succ),
                        std::memory_order_acq_rel,
                        std::memory_order_acquire);

                    succ = nodeToRemove->forward[level].load(
                        std::memory_order_acquire);
                }
            }

            Node* succ = nodeToRemove->forward[0].load(
                std::memory_order_acquire);

            while (true) {
                if (is_marked_ref(succ)) {
                    isMarked = true;
                    break;
                }

                if (nodeToRemove->forward[0].compare_exchange_strong(
                        succ,
                        get_marked_ref(succ),
                        std::memory_order_acq_rel,
                        std::memory_order_acquire)) {
                    isMarked = true;
                    find(key, preds, succs);
                    return true;
                }
            }

            if (isMarked) {
                return false;
            }
        }
    }
};

File: generated_codes/chat_agents/chat-agents-java/skiplist/claude_java/output1_skiplist.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {

    private static final int MAX_LEVEL = 16;
    private static final double PROBABILITY = 0.5;

    private static class Node {
        final int key;
        final int level;
        final AtomicMarkableReference<Node>[] forward;

        @SuppressWarnings("unchecked")
        Node(int key, int level) {
            this.key = key;
            this.level = level;
            this.forward = new AtomicMarkableReference[level + 1];
            for (int i = 0; i <= level; i++) {
                this.forward[i] = new AtomicMarkableReference<>(null, false);
            }
        }
    }

    private final Node head;
    private final Node tail;

    @SuppressWarnings("unchecked")
    public ConcurrentDataStructure() {
        head = new Node(Integer.MIN_VALUE, MAX_LEVEL);
        tail = new Node(Integer.MAX_VALUE, MAX_LEVEL);
        for (int i = 0; i <= MAX_LEVEL; i++) {
            head.forward[i].set(tail, false);
        }
    }

    private int randomLevel() {
        int lvl = 0;
        while (lvl < MAX_LEVEL && Math.random() < PROBABILITY) {
            lvl++;
        }
        return lvl;
    }

    @SuppressWarnings("unchecked")
    private Node[] find(int key, Node[] preds, Node[] succs) {
        boolean[] marked = {false};
        boolean snip;
        Node pred, curr, succ;

        retry:
        while (true) {
            pred = head;
            for (int level = MAX_LEVEL; level >= 0; level--) {
                curr = pred.forward[level].getReference();
                while (true) {
                    succ = curr.forward[level].get(marked);
                    while (marked[0]) {
                        snip = pred.forward[level].compareAndSet(curr, succ, false, false);
                        if (!snip) continue retry;
                        curr = pred.forward[level].getReference();
                        succ = curr.forward[level].get(marked);
                    }
                    if (curr.key < key) {
                        pred = curr;
                        curr = succ;
                    } else {
                        break;
                    }
                }
                preds[level] = pred;
                succs[level] = curr;
            }
            return succs;
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public boolean add(int key) {
        int topLevel = randomLevel();
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];

        while (true) {
            find(key, preds, succs);
            if (succs[0] != null && succs[0].key == key) {
                return false;
            }
            Node newNode = new Node(key, topLevel);
            for (int level = 0; level <= topLevel; level++) {
                newNode.forward[level].set(succs[level], false);
            }
            Node pred = preds[0];
            Node succ = succs[0];
            if (!pred.forward[0].compareAndSet(succ, newNode, false, false)) {
                continue;
            }
            for (int level = 1; level <= topLevel; level++) {
                while (true) {
                    pred = preds[level];
                    succ = succs[level];
                    if (pred.forward[level].compareAndSet(succ, newNode, false, false)) {
                        break;
                    }
                    find(key, preds, succs);
                }
            }
            return true;
        }
    }

    @Override
    public boolean remove(int key) {
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];
        boolean[] marked = {false};

        while (true) {
            find(key, preds, succs);
            Node nodeToRemove = succs[0];
            if (nodeToRemove == null || nodeToRemove.key != key) {
                return false;
            }
            for (int level = nodeToRemove.level; level >= 1; level--) {
                Node succ = nodeToRemove.forward[level].get(marked);
                while (!marked[0]) {
                    nodeToRemove.forward[level].compareAndSet(succ, succ, false, true);
                    succ = nodeToRemove.forward[level].get(marked);
                }
            }
            Node succ = nodeToRemove.forward[0].get(marked);
            while (true) {
                boolean iMarkedIt = nodeToRemove.forward[0].compareAndSet(succ, succ, false, true);
                if (iMarkedIt) {
                    // Node has been marked
                    find(key, preds, succs);
                    return true;
                } else {
                    succ = nodeToRemove.forward[0].get(marked);
                    if (marked[0]) {
                        return false;
                    }
                }
            }
        }
    }

    @Override
    public boolean contains(int key) {
        boolean[] marked = {false};
        Node curr = head;
        for (int level = MAX_LEVEL; level >= 0; level--) {
            Node next = curr.forward[level].getReference();
            while (next.key < key) {
                curr = next;
                next = curr.forward[level].getReference();
            }
        }
        curr = curr.forward[0].getReference();
        curr.forward[0].get(marked);
        return curr.key == key && !marked[0];
    }
}